File: overlay/python/hydra-kernel/src/main/python/hydra/dsl/meta/lib/equality.py
"""Phantom-typed term DSL for the hydra.lib.equality library."""

from hydra.typed import TypedTerm
from hydra.dsl.meta.phantoms import lib_primitive1 as primitive1, lib_primitive2 as primitive2


def compare(x: TypedTerm, y: TypedTerm) -> TypedTerm:
    """Compare two values and return a Comparison."""
    return primitive2(x, y)


def equal(a: TypedTerm, b: TypedTerm) -> TypedTerm:
    """Check if two values are equal."""
    return primitive2(a, b)


def gt(a: TypedTerm, b: TypedTerm) -> TypedTerm:
    """Check if first value is greater than second."""
    return primitive2(a, b)


def gte(a: TypedTerm, b: TypedTerm) -> TypedTerm:
    """Check if first value is greater than or equal to second."""
    return primitive2(a, b)


def identity(a: TypedTerm) -> TypedTerm:
    """Return the identity of a value."""
    return primitive1(a)


def lt(a: TypedTerm, b: TypedTerm) -> TypedTerm:
    """Check if first value is less than second."""
    return primitive2(a, b)


def lte(a: TypedTerm, b: TypedTerm) -> TypedTerm:
    """Check if first value is less than or equal to second."""
    return primitive2(a, b)


def max_(a: TypedTerm, b: TypedTerm) -> TypedTerm:
    """Return the maximum of two values."""
    return primitive2(a, b)


def min_(a: TypedTerm, b: TypedTerm) -> TypedTerm:
    """Return the minimum of two values."""
    return primitive2(a, b)

File: overlay/haskell/hydra-kernel/src/main/haskell/Hydra/Dsl/Meta/Lib/Equality.hs
-- | Phantom-typed term DSL for the hydra.lib.equality library

{-# LANGUAGE FlexibleContexts #-}

module Hydra.Dsl.Meta.Lib.Equality where

import Hydra.Dsl.AsTerm
import Hydra.Dsl.Meta.Phantoms
import Hydra.Typed
import Hydra.Util
import qualified Hydra.Dsl.Terms as Terms

import Data.Int
import qualified Hydra.Dsl.Prims as Prims
import qualified Hydra.Lib.Equality as DefEquality


-- | Compare two values and return a Comparison.
compare :: AsTerm t2 a => TypedTerm a -> t2 -> TypedTerm Comparison
compare t1 t2 = primitive2 DefEquality.compare t1 (asTerm t2)

-- | Check if two values are equal.
equal :: AsTerm t2 a => TypedTerm a -> t2 -> TypedTerm Bool
equal t1 t2 = primitive2 DefEquality.equal t1 (asTerm t2)

-- | Check if first value is greater than second.
gt :: AsTerm t2 a => TypedTerm a -> t2 -> TypedTerm Bool
gt t1 t2 = primitive2 DefEquality.gt t1 (asTerm t2)

-- | Check if first value is greater than or equal to second.
gte :: AsTerm t2 a => TypedTerm a -> t2 -> TypedTerm Bool
gte t1 t2 = primitive2 DefEquality.gte t1 (asTerm t2)

-- | Return a value unchanged.
identity :: TypedTerm a -> TypedTerm a
identity = primitive1 DefEquality.identity

-- | Check if first value is less than second.
lt :: AsTerm t2 a => TypedTerm a -> t2 -> TypedTerm Bool
lt t1 t2 = primitive2 DefEquality.lt t1 (asTerm t2)

-- | Check if first value is less than or equal to second.
lte :: AsTerm t2 a => TypedTerm a -> t2 -> TypedTerm Bool
lte t1 t2 = primitive2 DefEquality.lte t1 (asTerm t2)

-- | Return the maximum of two values.
max :: AsTerm t2 a => TypedTerm a -> t2 -> TypedTerm a
max t1 t2 = primitive2 DefEquality.max t1 (asTerm t2)

-- | Return the minimum of two values.
min :: AsTerm t2 a => TypedTerm a -> t2 -> TypedTerm a
min t1 t2 = primitive2 DefEquality.min t1 (asTerm t2)
